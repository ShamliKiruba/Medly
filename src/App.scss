.body {
    margin: 50px 100px;
    h2 {
        text-align: center;
    }
    .part-one {
        display: flex;
        .slider-component {
            .container {
                padding: 32px;
                .display {
                    margin-bottom: 10px;
                }
            }
            input[type=range] {
                width: 500px;
            }
            input[type=text] {
                padding: 10px;
				border-radius: 3px;
				border: 1px solid #b7b4b4;
				font-size: 15px;
            }
            .symbol {
                display: inline;
                button {
                    border: none;
					background: transparent;
					font-size: 15px;
                    height: 100%;
                    padding: 10px;
                    border: 1px solid #c3c3c3;
                    background: #e9ecef;
                    &:focus {
                        border: 1px solid grey;
                    }
                }
                button:focus{
                    outline: none;
                }
                #Yr, #Mon {
                    cursor: pointer;
                    &:hover {
                        background-color: #c3c3c3;
                    }
                }
                .selectedUnit {
                    background-color: #c3c3c3;
                    border: 1px solid grey;
                }
            }
            .marker {
                width: 500px;
                margin: 5px;
                display: flex;
                position: relative;
                div {
                    position: absolute;
                }
            }
        }
        .summary {
            padding: 2%;
            margin: auto;
            display: flex;
            border: 1px dotted #8b8c8e;
            justify-content: space-around;
            .data {
                display: flex;
                flex-direction: column;
                .label {
                    font-weight: bold;
                }
            }
            #chart {
                .highcharts-credits {
                    display: none;
                }
            }
        }
    }
    .part-two {
        .date {
            display: flex;
            margin: 5%;
            justify-content: center;
            p {
                margin: 5px 15px;
            }
        }
        table, td, th {  
            border: 1px solid #ddd;
            text-align: left;
        }
    
        table {
            border-collapse: collapse;
            width: 100%;
            table-layout: fixed;
            th, td {
                padding: 15px;
            }
            tr {
                table-layout: fixed;
            }
            .expand-collapse {
                cursor: pointer;
                &:hover{
                    background-color: aliceblue;
                }
            }
            .selectedRow {
                background-color: #b3b6b9;
            }
            .hide {
                display: none;
            }
            .show {
                background-color: #d0d0d0;
                display: table-row;
            }
        }	
    }
}


