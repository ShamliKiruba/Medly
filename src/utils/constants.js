export const LOAN_DETAILS = [
    {
        label: 'Loan EMI',
        key: 'emi',
    },
    {
        label: 'Total Interest Payable',
        key: 'totalInterest',
    },
    {
        label: 'Total Payment',
        key: 'totalPayment',
    }
];

export const MONTH_LIST = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

export const TABLE_TITLE = ['Year', 'Principal (A)', 'Interest (B)', 'Total Payment (A + B)', 'Balance', 'Loan Pain To Date'];

export const INIT_TABLEDATA = {
    thead: TABLE_TITLE,
    tbody: [{
        year: new Date().getFullYear(),
        month: '',
        principal: 1,
        interest: 1,
        total: 1,
        balance: 1,
        loanPaid: 1
    }],
};

export const INIT_DATE = {
    date: (new Date().getDate()) < 10 ? `0${(new Date().getDate())}`: `${new Date().getDate()}`,
    month: (new Date().getMonth() + 1) < 10 ? `0${(new Date().getMonth() + 1)}`: `${new Date().getMonth() + 1}`,
    year: `${new Date().getFullYear()}`,
};

export const INIT_LOAN_INFO = {
    loanAmount: 5000000,
    interestRate: 10.5,
    loanTenure: 20
};

export const INIT_ANNUAL_INFO = {
    emi: 0,
    totalPayment: 0,
    totalInterest: 0,
};