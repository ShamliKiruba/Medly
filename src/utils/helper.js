import { MONTH_LIST, TABLE_TITLE } from './constants';

export const range = (limit, step, spacing, unit) => {
    let marks = [];
    for(let i = 0; i <= limit; i = i+step) {
        let obj = {
            value: i,
            label: unit ? (i+unit) : i,
            spacing
        }
        marks.push(obj);
    }
    return marks;
};

export const LOAN_AMOUNT_MARKER = range(200, 25, 60, 'L');
export const INTEREST_RATE_MARKER = range(20, 2.5, 60);
export const TENURE_YEAR_MARKER = range(30, 5, 80);
export const TENURE_MONTH_MARKER = range(360, 60, 80);

export const TENURE_MONTH = {
    label: 'Loan Tenure',
    id:'loanTenure',
    defaultValue: 20,
    step: 6,
    min: 0,
    max: 360,
    marks: TENURE_MONTH_MARKER,
    selectedUnit: 'Mon',
    units: ['Yr', 'Mon']
};

export const TENURE_YEAR = {
    label: 'Loan Tenure',
    id:'loanTenure',
    defaultValue: 20,
    step: 0.5,
    min: 0,
    max: 30,
    selectedUnit: 'Yr',
    marks: TENURE_YEAR_MARKER,
    units: ['Yr', 'Mon']
};

export const RANGE_SLIDER = [
    {
        label: 'Home Loan Amount',
        id: 'loanAmount',
        defaultValue: 25,
        step: 50000,
        min: 0,
        max: 20000000,
        marks: LOAN_AMOUNT_MARKER,
        units: ['₹']
    },
    {
        label: 'Interest Rate',
        id:'interestRate',
        defaultValue: 10.5,
        step: 0.5,
        min: 0,
        max: 20,
        marks: INTEREST_RATE_MARKER,
        units: ['%']
    },
    TENURE_YEAR,
];

export const calculateEMI = (loanInfo, duration) => {
    let interest = ((loanInfo.interestRate /12) * 0.01);
    let timeDuration = duration === 'Yr' ? (loanInfo.loanTenure * 12) : (loanInfo.loanTenure)
    let a = Math.pow((1 + interest), timeDuration)
    let b = a - 1;
    let emi = loanInfo.loanAmount * (interest) * (a/b);
    let totalPayment =  emi * timeDuration;
    let totalInterest = (totalPayment - loanInfo.loanAmount);
    return {
        emi: Math.round(emi),
        totalPayment: Math.round(totalPayment),
        totalInterest: Math.round(totalInterest),
    }
};

export const createTable = (props, dateObj) => {
    let durationInMonths = props.duration === 'Yr' ? props.loanInfo.loanTenure * 12 : props.loanInfo.loanTenure;
    let sumA = 0, sumB = 0, sumC = 0, sumD = 0, sumE = 0, parent = {}, arr = [], parentArr = [], year = Number(dateObj.year);
    let emi = Number(props.annualInfo.emi);
    let month = Number(dateObj.month);
    let balanceAmount = props.loanInfo.loanAmount, sumOfPrincipal = 0;
    let initialAmount = props.loanInfo.loanAmount;
    let childAccumulator = [];
    for(let i=0; i<durationInMonths;i++) {
        let temp = {
            year: year,
            month: MONTH_LIST[month-1],
            total: emi,
            loanPaid:  0
        };
        let interestComponent = (props.loanInfo.interestRate * balanceAmount)/1200;
        let principalComponent = emi - interestComponent;
        temp.balance = balanceAmount - principalComponent; // final
        temp.interest = interestComponent;
        temp.principal = principalComponent;

        balanceAmount = temp.balance;
        sumOfPrincipal+= temp.principal;
        temp.loanPaid = ((sumOfPrincipal / initialAmount) * 100).toFixed(2);

        sumA+=temp.principal;
        sumB+=temp.interest;
        sumC+=temp.total;
        
        childAccumulator.push({
            month: temp.month,
            principal: `₹ ${convertValue(Math.round(temp.principal))}`,
            interest: `₹ ${convertValue(Math.round(temp.interest))}`,
            total: `₹ ${convertValue(emi)}`,
            balance: `₹ ${convertValue(Math.round(temp.balance))}`,
            loanPaid: `${temp.loanPaid} %`,
            year: temp.year,
        });
        if(MONTH_LIST[month-1] === 'Dec'  || (i === durationInMonths - 1)) {
            sumD=balanceAmount;
            sumE+=Number(temp.loanPaid);
            parent = {
                year, 
                sumA: `₹ ${convertValue(Math.round(sumA))}`, 
                sumB: `₹ ${convertValue(Math.round(sumB))}`, 
                sumD: (i === durationInMonths - 1) ? 0 : `₹ ${convertValue(Math.round(sumD))}`, 
                sumC: `₹ ${convertValue(sumC)}`,
                sumE: `${sumE} %`
            };
            arr.push(parent, ...childAccumulator);
            parentArr.push({
                year, 
                sumA: Math.round(sumA), 
                sumB: Math.round(sumB), 
                sumD: (i === durationInMonths - 1) ? 0 : Math.round(sumD), 
                sumC, 
                sumE
            });
            year++;
            sumA = 0; sumB = 0; sumC = 0; sumE = 0;
            month = 1;
            childAccumulator = [];
        } else {
            month++;
        }
    }
    return { tableData: {
        thead: TABLE_TITLE,
        tbody: arr
    }, parentArr };
};

export const convertValue = (value = 0) => {
    return value.toLocaleString('en-IN');
}