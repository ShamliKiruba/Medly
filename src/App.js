import React, { useState, useEffect, useCallback } from 'react';
import { calculateEMI } from './utils/helper';
import { INIT_LOAN_INFO, INIT_ANNUAL_INFO } from './utils/constants';
import { updatePieChart } from './components/common/Chart';
import EMICalculation from './components/EMICalculation/EMICalculation';
import PaymentBreakdown from './components/PaymentBreakdown/PaymentBreakdown';

const App = () => {
    const [loanInfo, updateLoanInfo] = useState(INIT_LOAN_INFO);
    const [annualInfo, setAnnualInfo] = useState(INIT_ANNUAL_INFO);
    const [duration, updateDuration] = useState('Yr');

    const updateEMI = useCallback(() => {
        let updatedEMI = calculateEMI(loanInfo, duration);
        setAnnualInfo(updatedEMI);
		updatePieChart(updatedEMI.totalInterest, loanInfo.loanAmount);
	}, [duration, loanInfo]);

    useEffect(() => {
        updateEMI();
    }, [loanInfo, updateEMI]);

    const setData = (id, value) => {
        updateLoanInfo(prevState => ({
            ...prevState,
            [id]: value,
        }));
        updateEMI();
    };

    return (
        <div className="body">
            <h2>EMI CALCULATOR</h2>
            <EMICalculation
                duration={duration}
                loanInfo={loanInfo}
                annualInfo={annualInfo}
                setData={setData}
                updateDuration={updateDuration}
            />
            <PaymentBreakdown
                duration={duration}
                loanInfo={loanInfo}
                annualInfo={annualInfo}
            />
        </div>
    )
};

export default App;
