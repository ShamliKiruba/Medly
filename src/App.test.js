import React from 'react';
import { fireEvent,render } from '@testing-library/react';
import App from './App';
import {calculateEMI } from '../src/utils/helper'

describe('App', () => {
	it('should render App', () => {
		const { getByText } = render(<App />);
		const header = getByText('EMI CALCULATOR');
		const sliderElement = getByText('Interest Rate');
		expect(header).toBeInTheDocument();
		expect(sliderElement).toMatchSnapshot();
	});
	
	it('should calculate emi based on loan details', () => {
		const loanInfo = {
			loanAmount: 5000000,
			interestRate: 10.5,
			loanTenure: 20
		};
		const duration = 'Yr';
		const result = calculateEMI(loanInfo, duration);
	
		expect(result).toBeInstanceOf(Object);
		expect(result).toEqual({
			"emi": 49919,
			"totalInterest": 6980559,
			"totalPayment": 11980559
		});
	});
	
	it('should calculate loan emi on loan amount change', () => {
		const { container } = render(<App />);
		fireEvent.change(container.querySelector('#loanAmount'), {
			target : { value : 100000 }
		});
		const elInput = container.querySelector('#label-0').innerHTML;
		expect(elInput).toContain('998');
	});
	
	it('should change duration scale on unit change', () => {
		const { container } = render(<App />);
		const month = container.querySelector('#Mon');
		fireEvent.click(month, {
			target : { value : 'Mon' }
		});
		const nodes = container.querySelectorAll('.marker');
		const last = nodes[nodes.length- 1];
		const className = month.classList[0];
		expect(last.innerHTML).toContain('360');
		expect(className).toEqual('selectedUnit');
	});
});
