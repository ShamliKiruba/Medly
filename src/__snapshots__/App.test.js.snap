// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render App 1`] = `
<label>
   
  Interest Rate
   
</label>
`;
